### PR TITLE
[dag] derive macro for enum variant conversions 

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR adds a derive macro that implements `From` traits for converting structs in enum to enum variants and vice-versa.

### CODE REVIEW

**Review by sasha8 — APPROVED** [review]
Black magic. Can we add some tests?

**Review by zekun000 — CHANGES_REQUESTED** [review]
I think the idiomatic way is to derive based on the enum instead of individual struct, I'd imagine it's like
```
#[derive(EnumMessage)]
enum DAGMessage
```
and based on the internal struct to do dispatches

**Comment by ibalajiarun** [comment]
> I think the idiomatic way is to derive based on the enum instead of individual struct, I'd imagine it's like
> 
> ```
> #[derive(EnumMessage)]
> enum DAGMessage
> ```
> 
> and based on the internal struct to do dispatches

yes, this is better. The only catch is all the structs should be in the same module as the enum right?

**Review by zekun000 — APPROVED** [review]
I think the types just need to be in the same crate not necessarily module (not foreign types), but I could be wrong. 

another nit: the name is not very representative, I'd call it EnumConversion. the same for the crate name, it's probably better to call it aptos-enum-conversion-derive

File: crates/aptos-enum-conversion-derive/tests/cases/no_fields.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Copyright © Aptos Foundation
+
+use aptos_enum_conversion_derive::EnumConversion;
+
+#[derive(EnumMessage)]
```
Comment (zekun000) [inline crates/aptos-enum-conversion-derive/tests/cases/no_fields.rs:5]: you need to update the tests lol

Comment (ibalajiarun) [reply crates/aptos-enum-conversion-derive/tests/cases/no_fields.rs]: good catch.

**Pull request merged**